controllers/policies/policyserver_controller.go

```diff
@@ -234,24 +237,47 @@ func (r *PolicyServerReconciler) SetupWithManager(mgr ctrl.Manager) error {
 	return err
 }
 
-func (r *PolicyServerReconciler) reconcileOrphanPolicies(policy *policiesv1alpha2.ClusterAdmissionPolicy) reconcile.Request {
-	if policy.DeletionTimestamp != nil {
+func (r *PolicyServerReconciler) reconcileOrphanPolicies(policy policiesv1alpha2.Policy) reconcile.Request {
+	if policy.GetDeletionTimestamp() != nil {
 		// policy not associated with PolicyServer, and scheduled
 		// for deletion, remove finalizer:
-		patch := policy.DeepCopy()
+		var patch policiesv1alpha2.Policy
+		policy.CopyInto(&patch)
 		controllerutil.RemoveFinalizer(patch, constants.KubewardenFinalizer)
 		err := r.Client.Patch(context.Background(), patch, client.MergeFrom(policy))
 		if err != nil {
-			r.Log.Error(err, "cannot remove finalizer from policy "+policy.Name)
+			r.Log.Error(err, "cannot remove finalizer from policy "+policy.GetName())
 		}
 		return ctrl.Request{}
 	}
 
-	policy.Status.PolicyStatus = policiesv1alpha2.ClusterAdmissionPolicyStatusUnschedulable
+	policy.SetStatus(policiesv1alpha2.PolicyStatusUnschedulable)
 	err := r.Reconciler.UpdateAdmissionPolicyStatus(context.Background(), policy)
 	if err != nil {
-		r.Log.Error(err, "cannot update status of policy "+policy.Name)
+		r.Log.Error(err, "cannot update status of policy "+policy.GetName())
 	}
-	r.Log.Info("policy " + policy.Name + " cannot be scheduled: no matching PolicyServer")
+	r.Log.Info("policy " + policy.GetName() + " cannot be scheduled: no matching PolicyServer")
 	return ctrl.Request{}
 }
+
+func (r *PolicyServerReconciler) watchCreateFunc(e event.CreateEvent, queue workqueue.RateLimitingInterface) {
+	queue.Add(setDefaultClusterAdmissionPolicyStatus(r, e.Object))
```

Comment:
Ah, I see what you are doing here now. Sorry. Does the `err := r.Client.Patch(context.Background(), patch, client.MergeFrom(policy))` do the right thing when a `ClusterAdmissionPolicy` or an `AdmissionPolicy` are modified?

I guess the magic comes from `GetObjectMeta()` in the interface?

Reply:
In any case, I think it would be good to decouple the actions, like:

```
func (r *PolicyServerReconciler) watchPolicyCreation(e event.CreateEvent, queue workqueue.RateLimitingInterface) {
  // Get policy server name from event
  // Set default cluster admission policy status
  // Add policy server to the queue
```
```

The fact that `setDefaultClusterAdmissionPolicyStatus` returns the name of the policy server to be reconciled, so it's added to the queue seems a bit strange in my opinion.

Reply:
Changes in the status will not trigger an update event since we are using the [GenerationChangedPredicate event filter](https://github.com/raulcabello/kubewarden-controller/blob/main/controllers/policies/policyserver_controller.go#L220) 

> This predicate will skip update events that have no change in the object's metadata.generation field. The metadata.generation field of an object is incremented by the API server when writes are made to the spec field of an object. This allows a controller to ignore update events where the spec is unchanged, and only the metadata and/or status fields are changed.

Reply:
> Ah, I see what you are doing here now. Sorry. Does the `err := r.Client.Patch(context.Background(), patch, client.MergeFrom(policy))` do the right thing when a `ClusterAdmissionPolicy` or an `AdmissionPolicy` are modified?
> 
> I guess the magic comes from `GetObjectMeta()` in the interface?

This will apply a Patch for a Policy, which can be either a `ClusterAdmissionPolicy` or an `AdmissionPolicy`

Reply:
> Changes in the status will not trigger an update event since ...

Oh, that's totally fine. Sorry, I wasn't clear. What I meant is, since we are using the interface when calling to `Patch`, is the library able to properly build the URL to where to send the patch to, if the event is received by a `AdmissionPolicy` or a `ClusterAdmissionPolicy`?

This is what I meant that I guess it's, and the magic behind it being the `GetObject()` implementation, so the path built by controller-runtime is always fine -- regardless of the event that triggered this callback: `AdmissionPolicy` or `ClusterAdmissionPolicy` --.

The reason of my question, is because the same callbacks are shared for `ClusterAdmissionPolicy` and `AdmissionPolicy`: https://github.com/kubewarden/kubewarden-controller/pull/177/files#diff-eff894e5ab7a87938c01ec9989a2ac60c2c6b4989a2c06903be5e060ab0e91a7R222-R230

Reply:
Yes. Patch is expecting a client.Object, and all Polices also implement the [client.Object interface](https://github.com/raulcabello/kubewarden-controller/blob/main/apis/policies/v1alpha2/policy.go#L12) 
Therefore the controller-runtime will build the right request using the client.Object methods

Reply:
> In any case, I think it would be good to decouple the actions, like:
> 
> ```
> func (r *PolicyServerReconciler) watchCreateFunc(e event.CreateEvent, queue workqueue.RateLimitingInterface) {
>   // Get policy server name from event
>   // Set default cluster admission policy status
>   // Add policy server to the queue
> }
> ```
> 
> The fact that `setDefaultClusterAdmissionPolicyStatus` returns the name of the policy server to be reconciled, so it's added to the queue seems a bit strange in my opinion.

done in https://github.com/raulcabello/kubewarden-controller/commit/9b665d0a38852284b6e808708debc57ab0499c96